packages/utils/src/ast-utils/helpers.ts

```diff
@@ -1,38 +1,40 @@
 import { AST_NODE_TYPES, AST_TOKEN_TYPES, TSESTree } from '../ts-estree';
 
-type ObjectEntry<BaseType> = [keyof BaseType, BaseType[keyof BaseType]];
+type ObjectEntry<BaseType> = BaseType extends unknown
+  ? [keyof BaseType, BaseType[keyof BaseType]]
+  : never;
 type ObjectEntries<BaseType> = Array<ObjectEntry<BaseType>>;
 
 export const isNodeOfType =
   <NodeType extends AST_NODE_TYPES>(nodeType: NodeType) =>
   (
     node: TSESTree.Node | null | undefined,
-  ): node is TSESTree.Node & { type: NodeType } =>
+  ): node is Extract<TSESTree.Node, { type: NodeType }> =>
     node?.type === nodeType;
 
 export const isNodeOfTypes =
   <NodeTypes extends readonly AST_NODE_TYPES[]>(nodeTypes: NodeTypes) =>
   (
     node: TSESTree.Node | null | undefined,
-  ): node is TSESTree.Node & { type: NodeTypes[number] } =>
+  ): node is Extract<TSESTree.Node, { type: NodeTypes[number] }> =>
```

Comment:
If it's possible to move `Extract<TSESTree.Token, { type: TokenType }>` to another generic type, we can calculate it once (now – thrice, I doubt it caches the result) and it can be more performant 

e.g. `ExtractedNode` in https://tsplay.dev/mAKKXw

WDYT?

Same in other places

Reply:
This will make the code more readable as well imo, so good call! 👍